firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java

```diff
@@ -254,10 +264,19 @@ void recalculateAndSaveOverlays(Set<DocumentKey> documentKeys) {
   /** Queries the remote documents and overlays by doing a full collection scan. */
   private ImmutableSortedMap<DocumentKey, Document> getDocumentsMatchingCollectionQuery(
       Query query, IndexOffset offset) {
-    Map<DocumentKey, MutableDocument> remoteDocuments =
-        remoteDocumentCache.getAll(query.getPath(), offset);
     Map<DocumentKey, Mutation> overlays = documentOverlayCache.getOverlays(query.getPath(), -1);
 
+    // Remote document elision: avoid reading remote documents if their overlay is Set or Delete.
+    Set<DocumentKey> ignore = new HashSet<>();
+    for (Map.Entry<DocumentKey, Mutation> entry : overlays.entrySet()) {
+      if (entry.getValue() instanceof SetMutation || entry.getValue() instanceof DeleteMutation) {
+        ignore.add(entry.getKey());
+      }
+    }
+
+    Map<DocumentKey, MutableDocument> remoteDocuments =
+        remoteDocumentCache.getAll(query.getPath(), offset, ignore);
```

Comment:
I don't think this part of the optimization is quite worth it. We are still reading the elements from the cache. What do you think about not touching `getAll()`?

Reply:
It might be useful when user goes offline long enough and there are a lot documents with overlays. It is hard to say at what point it is worth it, but it does not hurt performance so let's keep it?

Reply:
I'd rather keep it as is. The old way is shorter, easier to understand and does not require porting. The only thing you are saving here is the Proto decoding, which already happens in the background.